### PR TITLE
[Functional Collectives] Migrate DeviceMesh::all_reduce to use functional all_reduce.

### PR DESCRIPTION
**Pull request by kumpera**
BC: This changes the signature and semantics of DeviceMesh::all_reduce.

DeviceMesh::all_reduce now uses a functional collective under the hood which makes it more easily traceable.
You no longer need to use CommTensor to get a trace.

all_reduce now is async only and uses AsyncCollectiveTensor to ensure proper stream synchronization.

Signature changed: removed `async_op` param and changes return type from `Optional[Work]` to `torch.Tensor`.

### CODE REVIEW

File: torch/distributed/_tensor/device_mesh.py
Comment (wanchaol) [inline torch/distributed/_tensor/device_mesh.py:435]: would this be AsyncTensor or just Tensor?

**Review by wanchaol — APPROVED** [review]
lgtm, thanks! quick stamp to unblock, please make sure CI pass before landing :)

File: test/distributed/_tensor/test_device_mesh.py
Comment (anj-s) [inline test/distributed/_tensor/test_device_mesh.py:243]: nit: type s/plan/plain

File: torch/distributed/_spmd/distribute.py
Comment (anj-s) [inline torch/distributed/_spmd/distribute.py:252]: if we move to using the new collectives will we need to check for "wait_comm"?

File: torch/distributed/_tensor/device_mesh.py
Comment (kumpera) [reply torch/distributed/_tensor/device_mesh.py:435]: It will be AsyncTensor in eager and plain tensor when tracing. Right @wconstab ?

Either way, do we want to document this fact?

File: torch/distributed/_spmd/distribute.py
Comment (kumpera) [reply torch/distributed/_spmd/distribute.py:252]: Once we move to the new collectives we'd stop using CommTensor in DMesh and we could drop supporting it.

This will require implementing all remaining collectives used by DMesh/DTensor which will take a while.

File: test/distributed/_tensor/test_device_mesh.py
Comment (kumpera) [inline test/distributed/_tensor/test_device_mesh.py:244]: note to self, we can multiply by 1 as well.

**Pull request closed without merging**
